torch_xla/experimental/scan.py

```diff
@@ -253,24 +263,92 @@ def fn_no_output_aliasing(*args):
     flat_x, _ = tree_flatten(x)
-    out = fwd_graph(*flat_carry, *flat_x)
+    with xp.Trace('aot_forward'):
+      out = fwd_graph(*flat_carry, *flat_x)
     actual_out, activations = split(out, num_out)
     carry, y = unflatten_fwd_out(actual_out)
     y = (y, activations)
     return carry, y
 
+  def forward(carry, x):
+    carry, (y, activations) = forward_core(carry, x)
+
+    # Remove activations that alias to inputs. Those will be added back
+    # in `alias_input`.
+    partial_activations = tuple(
+        v for i, v in enumerate(activations) if i not in aliased_activations)
+
+    y = (y, partial_activations)
+    return carry, y
+
+  def alias_input(partial_activations, xs):
+    """
+    Add back activations that are aliases to input tensors.
+
+    In principle, we could have `forward` return all the intermediate activations,
+    including those that are aliases to an input tensor. However, those inputs will
+    then be duplicated as part of the output of a `scan` call, because we want to
+    save all activations during the forward pass of a `scan`. The XLA compiler can't
+    optimize away this duplication likely because they're behind a DynamicSlice +
+    DynamicUpdateSlice, so we end up doubling the memory usage from those inputs.
+
+    To reduce memory usage, we can have `forward` return the activations that
+    don't alias to inputs, called `partial_activations`. The autograd implementation
+    of `scan` will call `alias_input` to add back activations that are aliases
+    of input tensors outside of a scan, turning the partial activations back to
+    full activations.
+    """
+    activations = list(partial_activations)
+    aliased_inputs = [
+        v for i, v in enumerate(tree_iter(xs)) if i in input_activation_aliases
+    ]
+    for (i, activation_idx) in enumerate(input_activation_aliases.values()):
+      activations.insert(activation_idx, aliased_inputs[i])
+    return tuple(activations)
+
   def backward(carry, x):
     grad_new_carry, _ = tree_flatten(carry)
     (grad_y, activations) = x
     grad_y, _ = tree_flatten_none(grad_y)
-    out = bwd_graph(*activations, *grad_new_carry, *grad_y)
+    with xp.Trace('aot_backward'):
+      out = bwd_graph(*activations, *grad_new_carry, *grad_y)
     grad_carry, grad_x = unflatten_bwd_out(out)
     return grad_carry, grad_x
 
-  return forward, backward
+  return forward, alias_input, backward
+
+
+def _find_input_activation_aliases(fake_carry_pytree, fake_x_pytree, num_out,
+                                   fwd_graph):
+  """
+  Find which activations are aliases to input tensors.
+
+  Returns:
+
+    A mapping from index into the flatttened
+    input pytree to the index into the list of intermediate activations.
+
+  """
+  flat_carry, _ = tree_flatten(fake_carry_pytree)
+  flat_x, _ = tree_flatten(fake_x_pytree)
+  _actual_out, activations = split(fwd_graph(*flat_carry, *flat_x), num_out)
+  input_ids = (id(v) for v in tree_iter(flat_x))
```

Comment:
probably inline those 2?

generator expressions can only be iterated once, so once you assign to a variable, you have to be careful to not use it twice (you didn't). But that is a mental burden to the reader.

Reply:
TIL. Done.